drv/sprot-api/src/lib.rs

```diff
@@ -3,64 +3,266 @@
-// result in a retry on the SP side and an ErrorRsp on the RoT side.
-// Note: Variant order must be maintained to stay compatible due to SP/RoT
-// version skew during update.
+use sprockets_common::msgs::{
+    RotError as SprocketsError, RotRequestV1 as SprocketsReq,
+    RotResponseV1 as SprocketsRsp,
+};
+use static_assertions::const_assert;
+use userlib::sys_send;
+
+const CRC16: Crc<u16> = Crc::<u16>::new(&CRC_16_XMODEM);
+pub const CRC_SIZE: usize = <u16 as SerializedSize>::MAX_SIZE;
+// XXX ROT FIFO size should be discovered.
+pub const ROT_FIFO_SIZE: usize = 8;
+pub const MAX_BLOB_SIZE: usize = 512;
+pub const MAX_REQUEST_SIZE: usize =
+    Header::MAX_SIZE + ReqBody::MAX_SIZE + MAX_BLOB_SIZE;
+pub const MAX_RESPONSE_SIZE: usize =
+    Header::MAX_SIZE + RspBody::MAX_SIZE + MAX_BLOB_SIZE;
+
+// For simplicity we want to be able to retrieve the header
+// in a maximum of 1 FIFO size read.
+const_assert!(Header::MAX_SIZE <= ROT_FIFO_SIZE);
+
+pub type Request = Msg<ReqBody>;
+pub type Response = Msg<Result<RspBody, SprotError>>;
+
+/// A message header for a request or response
+///
+/// It's important that this header be kept fixed size by limiting the use
+/// of rust types. This allows us to assume that Header::MAX_SIZE is also the
+/// exact size of the header.
+#[derive(Serialize, Deserialize, SerializedSize)]
+pub struct Header {
+    pub protocol: Protocol,
+    pub body_size: u16,
+}
+
+impl Header {
+    fn new(body_size: u16) -> Header {
+        Header {
+            protocol: Protocol::V2,
+            body_size,
+        }
+    }
+}
+
+/// Information about optional blobs appended to the body of messages
+/// before the CRC.
+pub struct BlobInfo {
+    pub offset: usize,
+    pub size: usize,
+}
+
+/// A sprot Msg that flows between the SP and RoT
+///
+/// The message is parameterized by a `ReqBody` or `RspBody`.
+///
+/// Note that `MSG`s do not implement `Serialize`, `Deserialize`, or
+/// `SerializedSize`, as they need to calculate and place a CRC in the buffer.
+/// `Msg`s sometimes include a an offset into the buffer where a binary blob
+/// resides.
+pub struct Msg<T> {
+    pub header: Header,
+    pub body: T,
+
+    // The index into the serialized buffer where an optional binary blob lives
+    pub blob: Option<BlobInfo>,
+}
+
+impl<T> Msg<T>
+where
+    T: Serialize + for<'a> Deserialize<'a> + SerializedSize,
+{
+    /// Serialize a `Header` followed by a `ReqBody` or `RspBody`, compute a CRC, serialize
+    /// the CRC, and return the total size of the serialized request.
+    pub fn pack(body: &T, buf: &mut [u8]) -> Result<usize, SprotProtocolError> {
+        // Serialize `body`
+        let mut size = hubpack::serialize(&mut buf[Header::MAX_SIZE..], body)?;
+
+        // Create a header, now that we know the size of the body
+        let header = Header::new(size.try_into().unwrap_lite());
+
+        // Serialize the header
+        size += hubpack::serialize(buf, &header)?;
```

Comment:
(If you buy my reasoning above, this also seems like a good use for `unwrap_lite`, since you've already panicked if buf is less than `Header::MAX_SIZE` above by slicing)